pkgs/by-name/ld/ld64/0007-Add-OpenSSL-based-CoreCrypto-digest-functions.patch

```diff
@@ -0,0 +1,311 @@
+From 36767c7345161baf0ab125f95c8557f8e24f25db Mon Sep 17 00:00:00 2001
+From: Randy Eckenrode <randy@largeandhighquality.com>
+Date: Tue, 9 Apr 2024 19:28:17 -0400
+Subject: [PATCH 7/8] Add OpenSSL-based CoreCrypto digest functions
+
+---
+ compat/CommonCrypto/CommonDigest.h    |  6 +++
+ compat/CommonCrypto/CommonDigestSPI.c | 21 +++++++++++
+ compat/CommonCrypto/CommonDigestSPI.h | 14 +++++++
+ compat/corecrypto/api_defines.h       | 10 +++++
+ compat/corecrypto/ccdigest.c          | 53 +++++++++++++++++++++++++++
+ compat/corecrypto/ccdigest.h          | 27 ++++++++++++++
+ compat/corecrypto/ccdigest_private.h  | 19 ++++++++++
+ compat/corecrypto/ccsha1.c            | 22 +++++++++++
+ compat/corecrypto/ccsha1.h            |  9 +++++
+ compat/corecrypto/ccsha2.c            | 22 +++++++++++
+ compat/corecrypto/ccsha2.h            |  9 +++++
+ 11 files changed, 212 insertions(+)
+ create mode 100644 compat/CommonCrypto/CommonDigest.h
+ create mode 100644 compat/CommonCrypto/CommonDigestSPI.c
+ create mode 100644 compat/CommonCrypto/CommonDigestSPI.h
+ create mode 100644 compat/corecrypto/api_defines.h
+ create mode 100644 compat/corecrypto/ccdigest.c
+ create mode 100644 compat/corecrypto/ccdigest.h
+ create mode 100644 compat/corecrypto/ccdigest_private.h
+ create mode 100644 compat/corecrypto/ccsha1.c
+ create mode 100644 compat/corecrypto/ccsha1.h
+ create mode 100644 compat/corecrypto/ccsha2.c
+ create mode 100644 compat/corecrypto/ccsha2.h
+
+diff --git a/compat/CommonCrypto/CommonDigest.h b/compat/CommonCrypto/CommonDigest.h
+new file mode 100644
+index 0000000..a60eba7
+--- /dev/null
++++ b/compat/CommonCrypto/CommonDigest.h
+@@ -0,0 +1,6 @@
++// SPDX-License-Identifier: APSL-2.0
++// CoreCrypto compatibility shims written by Randy Eckenrode © 2024
++
++#pragma once
++
++#define CCSHA256_OUTPUT_SIZE 32
+diff --git a/compat/CommonCrypto/CommonDigestSPI.c b/compat/CommonCrypto/CommonDigestSPI.c
+new file mode 100644
+index 0000000..41269fc
+--- /dev/null
++++ b/compat/CommonCrypto/CommonDigestSPI.c
+@@ -0,0 +1,21 @@
++// SPDX-License-Identifier: APSL-2.0
++// CoreCrypto compatibility shims written by Randy Eckenrode © 2024
++
++#include "CommonDigestSPI.h"
++
++#include <stdlib.h>
++#include <string.h>
++
++#include <corecrypto/ccsha2.h>
++
++void CCDigest(int type, const uint8_t* bytes, size_t count, uint8_t* digest) {
++    if (type != kCCDigestSHA256) {
++        abort();
++    }
++    const struct ccdigest_info* di = ccsha256_di();
++
++    ccdigest_di_decl(_di, ctx);
++    ccdigest_init(di, ctx);
++    ccdigest_update(di, ctx, count, bytes);
++    ccdigest_final(di, ctx, digest);
++}
```

Comment:
so this looks like it is the shim and only sha256 is supported -- there is code in the patch to do sha1 that seems unused, can that be deleted and then this patch can just contain one file and can get rid of the calls to malloc and the static variables and simplify this to just be:

```c
void CCDigest(int type, const uint8_t* bytes, size_t count, uint8_t* digest) {
    if (type != kCCDigestSHA256) {
        fprintf(stderr, "no digest implemeted for type %d\n", type);
        exit(1);
    }
    const EVP_MD* provider = EVP_sha256();
    EVP_MD_CTX *ctx = EVP_MD_CTX_new();

    bool success = provider != NULL
        && ctx != NULL
        && EVP_MD_get_size(provider) == 32
        && EVP_DigestInit_ex2(ctx, provider, NULL)
        && EVP_DigestUpdate(ctx, bytes, count)
        && EVP_DigestFinal_ex(ctx, digest, NULL);

    EVP_MD_CTX_free(ctx);

    if (!success) {
        ERR_print_errors_fp(stderr);
        exit(1);
    }
}
```

Reply:
There are two sets of shims: `CCDigest`, and `ccsha1_di` and `ccsha256_di`.

* Both `ccsha256_di` and `CCDigest` are used directly in [`OutputFile.cpp`](https://github.com/apple-oss-distributions/ld64/blob/47f477cb721755419018f7530038b272e9d0cdea/src/ld/OutputFile.cpp).
* `ccsha1_di` and `ccsha256_di` are referenced in [libcodedirectory.cpp](https://github.com/apple-oss-distributions/ld64/blob/47f477cb721755419018f7530038b272e9d0cdea/src/ld/libcodedirectory.c#L465-L471).

The goal is not to modify Apple’s code when using the shims (because that defeats the point), so dropping the `ccsha1_di` and `ccsha256_di` implementations is not really an option.


[1]: https://developer.apple.com/documentation/technotes/tn3126-inside-code-signing-hashes

Reply:
ah, bummer. i only saw them used in the patch.